online-shop/src/app/components/edit-product/edit-product.component.ts

```diff
@@ -0,0 +1,70 @@
+import {Component, OnInit} from '@angular/core';
+import {FormBuilder, FormGroup, Validators} from "@angular/forms";
+import {ActivatedRoute} from "@angular/router";
+import {HttpClient} from "@angular/common/http";
+import {ProductService} from "../../services/product.service";
+import {Location} from "@angular/common";
+import {ProductModel} from "../../types/product.model";
+
+
+@Component({
+  selector: 'app-edit-product',
+  templateUrl: './edit-product.component.html',
+  styleUrls: ['./edit-product.component.scss']
+})
+export class EditProductComponent implements OnInit {
+
+  product: ProductModel | undefined;
+  id: string | null = this.route.snapshot.paramMap.get('id');
+  form: FormGroup = this.fb.group({
+    name: ["the name", [Validators.required, Validators.minLength(5)]],
+    category: ["the category", [Validators.required]],
+    image: ["the image", [Validators.required]],
+    price: [0, [Validators.required]],
+    description: ["the description", [Validators.required]]
+  });
+
+  constructor(private fb: FormBuilder, private route: ActivatedRoute, private http: HttpClient,
+              private productService: ProductService, private location: Location) {
+  }
+
+  ngOnInit(): void {
+    if (this.id) {
+      this.productService.getProductByID(this.id).subscribe((data) => {
```

Comment:
**suggestion:** take care to avoid memory leaks in your code by unsubscribing from observables. https://www.thisdot.co/blog/best-practices-for-managing-rxjs-subscriptions/